text/0002-numeric-values.md

```diff
@@ -0,0 +1,485 @@
+
+## Parsing numbers
+
+For reference, the following regular expression is a set of all valid numeric literals:
+
+```regexp
+^(?:\d(?:\d|_\d)*)?\.?(?:\d(?:\d|_\d)*)+(?:e[+-]?(?:\d(?:\d|_\d)*))?\b
+```
+
+A compiler should implement this in native code than using a regular expression engine.
+
+Examples of valid numeric literals:
+
+- `10` (an integer)
+- `512_000.000` (an integer)
+- `63_2.42` (a float)
+- `.425_000` (a float)
+- `50.2e+3` (an integer)
+- `4_2.1_2e8_2` (an integer)
+
+Examples of invalid numeric literals:
+
+- `10.` (this dot may be seen as a property access)
+- `1.5.1` (potentially a property access)
+- `4e-4.0` (potentially a property access)
+- `_52` (an identifier)
+- `10_` (parse error: numeric separator must be between two numbers)
+- `1__3` (parse error: numeric separator must be between two numbers)
+
+## Type of numeric literals
+
+Numeric literals do not have explicit types when in expressions, but rather adapt to the runtime-known types. This works exactly how [Zig's `comptime_int` type](https://ziglang.org/documentation/master/#toc-Runtime-Integer-Values) works (the type of integer literals), but this RFC does not give this type an observable name.
+
+```
+var x: u32 = 512
+// `512` becomes a u32 based on the type of `x`
+```
+
+Literals that do not fit in the result type will cause a compile error.
+
+```
+var x: u32 = -2
+// compile error: cannot coerce `-2` to `u32`
+```
+
+You cannot use [unary `~`](#integer-arithmetic) on an integer literal, it must be given a type, such as:
+
+```blight
+~512 // compile error: cannot infer type of constant numeric literal
+     // hint: unary operator '~' needs a result type
+
+~512.to(u32) // ok
+```
+
+Similarly, runtime-mutable variables cannot be assigned without annotating their type.
+
+> [!NOTE]
+> There is not yet an RFC for variable declarations. We will assume `var name: type` is the syntax.
+
+```blight
+var x = 512 // compile error: cannot infer type of constant numeric literal
+
+var x: u32 = 512 // ok
+var x = 512.to(u32) // ok, less preferred approach
+```
+
+## Constant Expression Evaluation
+
+Expressions that contain only numeric literals must be evaluated at compile time.
+
+```blight
+var x: u32 = (2.5 + 2.5) / 5
+// valid as `1` can coerce to `u32`
+
+var x: u32 = 5 / 6
+// compile error: cannot coerce `0.8333333333333334` to `u32`
+```
+
+If undefined behavior is caused by a constant expression, it is a compile error.
+
+```blight
+var x: u32 = 5 / 0
+// compile error: division by zero
+```
+
+## Widening during arithmetic
+
+When integers of different sizes are used in arithmetic, the compiler will widen the smaller integer to the larger integer.
+
+For example, if a variable of type `u32` named `x` existed, and the following code was run:
+
+```blight
+x + 512
+```
+
+`512` is widened to a u32. If the rhs was larger than `u32.max`, then this would be a compile error. Explicit widening of `x` is needed.
```

Comment:
what is the rhs? The right hand side? This should be more explicit.

Reply:
will edit the document to clarify